File: backend/app/schemas/thermal.py
from datetime import UTC, datetime

from pydantic import BaseModel, ConfigDict, Field

from app.models import SessionStatus


def serialize_utc_datetime(value: datetime) -> str:
    normalized = value.replace(tzinfo=UTC) if value.tzinfo is None else value.astimezone(UTC)
    return normalized.isoformat().replace("+00:00", "Z")


class ReadModel(BaseModel):
    model_config = ConfigDict(from_attributes=True, json_encoders={datetime: serialize_utc_datetime})


class PrinterBase(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    base_url: str = Field(min_length=1, max_length=255)
    api_key: str | None = Field(default=None, max_length=255)
    notes: str | None = None
    is_enabled: bool = True


class PrinterCreate(PrinterBase):
    pass


class PrinterUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=120)
    base_url: str | None = Field(default=None, min_length=1, max_length=255)
    api_key: str | None = Field(default=None, max_length=255)
    notes: str | None = None
    is_enabled: bool | None = None


class PrinterRead(ReadModel, PrinterBase):
    id: int
    created_at: datetime
    updated_at: datetime


class PrinterConnectionCheckRead(BaseModel):
    printer_id: int
    reachable: bool
    status_code: int | None
    message: str
    moonraker_version: str | None
    klippy_state: str | None


class SessionStartRequest(BaseModel):
    label: str | None = Field(default=None, max_length=160)


class SessionStopRequest(BaseModel):
    stop_reason: str | None = Field(default=None, max_length=80)


class SessionDispositionRequest(BaseModel):
    save_notes: str | None = None


class SessionRead(ReadModel):
    id: int
    printer_id: int
    label: str | None
    started_at: datetime
    ended_at: datetime | None
    status: SessionStatus
    stop_reason: str | None
    save_notes: str | None
    sample_count: int = 0
    created_at: datetime
    updated_at: datetime


class TemperatureSampleRead(ReadModel):
    id: int
    session_id: int
    captured_at: datetime
    nozzle_actual: float | None
    nozzle_target: float | None
    bed_actual: float | None
    bed_target: float | None
    chamber_actual: float | None
    heater_power: float | None
    fan_speed: float | None
    print_state: str | None
    source: str
    raw_payload: str | None
    created_at: datetime
    updated_at: datetime


class ThermalEventRead(ReadModel):
    id: int
    session_id: int
    event_type: str
    message: str
    event_time: datetime
    metadata_json: str | None
    created_at: datetime
    updated_at: datetime


class SessionCaptureResponse(ReadModel):
    session: SessionRead
    sample: TemperatureSampleRead

File: frontend/src/types/thermal.ts
export type WatchRetentionHours = 4 | 8 | 12 | 24;
export type PreservedWatchCaptureStatus = "collecting" | "finalized";

export interface BackgroundWatchConfig {
  id: number;
  printer_id: number;
  is_enabled: boolean;
  retention_hours: WatchRetentionHours;
  poll_interval_seconds: number;
  created_at: string;
  updated_at: string;
}

export interface SmartWatchConfig {
  id: number;
  printer_id: number;
  is_enabled: boolean;
  last_observed_state: string | null;
  last_observed_filename: string | null;
  created_at: string;
  updated_at: string;
}

export interface PrinterProfile {
  id: number;
  name: string;
  base_url: string;
  api_key: string | null;
  notes: string | null;
  is_enabled: boolean;
  watch_config: BackgroundWatchConfig | null;
  smart_watch_config: SmartWatchConfig | null;
  created_at: string;
  updated_at: string;
}

export interface PrinterCreateInput {
  name: string;
  base_url: string;
  api_key?: string | null;
  notes?: string | null;
  is_enabled: boolean;
}

export interface PrinterConnectionCheck {
  printer_id: number;
  reachable: boolean;
  status_code: number | null;
  message: string;
  moonraker_version: string | null;
  klippy_state: string | null;
}

export interface BackgroundWatchConfigUpdate {
  is_enabled?: boolean;
  retention_hours?: WatchRetentionHours;
}

export interface SmartWatchConfigUpdate {
  is_enabled?: boolean;
}

export interface BackgroundWatchPromoteInput {
  label?: string | null;
  save_notes?: string | null;
  hours?: WatchRetentionHours | null;
}

export type SessionStatus = "active" | "completed" | "saved" | "discarded";
export type ComparisonAlignment = "elapsed" | "absolute";

export interface SessionRecord {
  id: number;
  printer_id: number;
  label: string | null;
  started_at: string;
  ended_at: string | null;
  status: SessionStatus;
  stop_reason: string | null;
  save_notes: string | null;
  smart_watch_run: SmartWatchRun | null;
  sample_count: number;
  created_at: string;
  updated_at: string;
}

export interface SmartWatchRun {
  id: number;
  printer_id: number;
  session_id: number;
  print_filename: string | null;
  started_state: string;
  last_state: string | null;
  terminal_state: string | null;
  started_via_recovery: boolean;
  created_at: string;
  updated_at: string;
}

export interface TemperatureTraceSample {
  id: number;
  captured_at: string;
  nozzle_actual: number | null;
  nozzle_target: number | null;
  bed_actual: number | null;
  bed_target: number | null;
  chamber_actual: number | null;
  heater_power: number | null;
  fan_speed: number | null;
  print_state: string | null;
  source: string;
  raw_payload: string | null;
  created_at: string;
  updated_at: string;
}

export interface TemperatureSample extends TemperatureTraceSample {
  session_id: number;
}

export interface BackgroundWatchSample extends TemperatureTraceSample {
  printer_id: number;
}

export interface PreservedWatchSample extends TemperatureTraceSample {
  capture_id: number;
  source_watch_sample_id: number | null;
}

export interface ThermalEvent {
  id: number;
  session_id: number;
  event_type: string;
  message: string;
  event_time: string;
  metadata_json: string | null;
  created_at: string;
  updated_at: string;
}

export interface PreservedWatchTriggerEvent {
  id: number;
  capture_id: number;
  event_time: string;
  trigger_rule: string;
  trigger_reason: string;
  metadata_json: string | null;
  created_at: string;
  updated_at: string;
}

export interface PreservedWatchCapture {
  id: number;
  printer_id: number;
  status: PreservedWatchCaptureStatus;
  source: string;
  trigger_rule: string;
  trigger_reason: string;
  trigger_time: string;
  capture_start_at: string;
  capture_end_at: string;
  finalized_at: string | null;
  sample_count: number;
  trigger_count: number;
  created_at: string;
  updated_at: string;
}

export interface SessionCaptureResponse {
  session: SessionRecord;
  sample: TemperatureSample;
}
